### ISSUE TITLE
Discussion on improvements around usermode hooking

### ISSUE THREAD

**Issue by icedevml** [tklengyel/drakvuf#669]
### What is not perfect about usermode hooks?
**Currently, they are equipped only when DRAKVUF starts.** The assumption is that there would be at least one program running in the system, which has the required DLL loaded. Moreover, the expectation is that for each program in the system, the function’s entry point will be backed by the same physical page, which is rather usually true, but doesn’t cover some special cases. We believe that hooks do work in socketmon just because it is lucky enough to always have some process in the system which is using socket APIs.

**Return hooks are somehow bugged.** Creating a hook on the first instruction after call to API function and then deleting it in callback makes the whole VM misbehave - the physical page that was used for return hook (that might be reused for later use, e.g. when running the application again) seems not to be writable anymore, which causes relocations to be corrupted. 

**The DLL is not always fully covered in memory after loading.** Because Windows is using some copy-on-write mechanism on DLLs, it may occur that some virtual pages inside the new DLL are backed by physical pages, but some of them not.

We tried to hook `NtMapViewOfSection` and `NtProtectVirtualMemory` syscalls in order to observe if some new DLL is being loaded into the process. Then, we tried to use the code similiar to the [socketmon’s module_trap_visitor](https://github.com/tklengyel/drakvuf/blob/48b36783de8ae65d267671efd777a33db6115179/src/plugins/socketmon/socketmon.cpp#L978) in order to equip hooks. Basically we have observed two interesting misbehaviours inside that function:

* for some DLLs, the PE header is readable, some export information is unreadable and thus `vmi_translate_sym2v` fails, e.g. with [“failed to get aof index”](https://github.com/tklengyel/libvmi/blob/bd7f212de8ad3d6e169e90209b86aa90bd880ca4/libvmi/os/windows/peparse.c#L534).
* for some DLLs, the PE header is readable, export information is readable, the entry point of the exported function is unreadable and thus [drakvuf_add_trap](https://github.com/tklengyel/drakvuf/blob/48b36783de8ae65d267671efd777a33db6115179/src/plugins/socketmon/socketmon.cpp#L1015) fails on translation from VA to PA.

More fun - hooking usually succeeds if we wait for the program to make a first call to the target function (e.g. `MessageBoxA`) and then try to use the abovementioned code. Perhaps the problems are due to this “lazy loading” of memory.

### What we tried to improve hooking?
**MEMACCESS breakpoints on PTEs.** Whenever we were stuck due to the fact that some virtual address is not readable, we created a `MEMACCESS` trap on the page table entry referring to this VA (and page directory, and so on…). This was working, but it was actually very slow. PoC-quality, extremely dirty code.

**Shellcode injection:** Whenever we were stuck, we were injecting a shellcode into the kernel which was “touching” the inaccessible memory addresses so we were able to equip hooks afterwards. This was also working, but it was kind-of invasive. PoC quality with code designed a little bit better than with other approaches.

**LibVMI Pagefault Injection:** After all, we saw that LibVMI added a new feature for page fault injection and used it instead of injecting shellcode. However, we are not doing it too cleanly in our PoC at the moment. If there is a problem with DLL then we are page-faulting all possible pages inside that DLL and then re-trying hooking because it was just a simple technique for a PoC. Very dirty code without a clear idea how to implement it in a reasonable way.

### Summary
It seems like the most reasonable way would be to implement some solution which:

* would observe new DLLs being loaded by Windows loader using hooks on syscalls (`NtProtectVirtualMemory` and `NtMapViewOfSection`)
* try to hook the required functions in DLLs
* on failure, do some page faults, but optimally (requires knowing which particular virtual addresses are required, which is not trivial)

The expected result would be development of a usermode hooking framework which would enhance existing plugins like socketmon and provide tools for memdump plugin.

What do you think about the overall design?

/cc @tklengyel @skvl @sasza8 @msm-code

**Comment by tklengyel** [comment]
From the options you tried I would say the pagefault injection is the most reasonable one. You probably don't have to pagefault all pages for the process, only the ones you need to find and add the traps you want.

I would also add to the list of problems is when pages get paged out, which could happen for monitored pages, and then when they are brought back they might not be at the same physical location. So there is a need for logic that identifies when that happens to either bring those pages back right away or to track and re-add breakpoints when the pages are back.

**Comment by icedevml** [comment]
@tklengyel We are experimenting with `MiCopyOnWrite` hook, in order to account for the situation when the program is overriding the DLL functions.

I guess there is also some routine for paging out which could be hooked and prevented from removing the required pages.

**Comment by icedevml** [comment]
@tklengyel First thing that we would like to investigate is:

> Return hooks are somehow bugged. Creating a hook on the first instruction after call to API function and then deleting it in callback makes the whole VM misbehave - the physical page that was used for return hook (that might be reused for later use, e.g. when running the application again) seems not to be writable anymore, which causes relocations to be corrupted.

Do you have any idea why such thing happens? We could provide some PoC if it's not clear enough.

**Comment by tklengyel** [comment]
Generally speaking dynamic traps are a PITA (adding traps in other event's callback and removing them in callbacks). That's why you don't see that being used very often. Doing that in an application (userspace) is just another layer of PITA. In this case since you say the application exits you have to make sure the traps you added are all removed when that happens, otherwise you might have a lingering trap when in fact its underlying memory is now being reused somewhere else.

**Comment by icedevml** [comment]
Another thing that I consider tricky is that we need to return from the hook function in which we call `vmi_request_page_fault` just to allow Xen to actually cause the page fault and repeat the previous instruction, so we are back in the beginning of our hook.

E.g.:
```cpp
void hook_NtDoSomething()
{
    if (address_is_accessible())
    {
        // TODO do stuff with the page
    }
    else if (!address_is_accessible() && first_attempt)
    {
        vmi_request_page_fault(vmi, info->vcpu, inaccessible_vaddr, 0);
    }
    else
    {
        PRINT_DEBUG("Still cannot access memory after causing page fault, giving up");
    }

    return VMI_EVENT_RESPONSE_NONE;
}
```

The code starts looking pretty ugly when we have some more advanced logic around call to `vmi_request_page_fault`. It's a textbook example of the place where coroutines would work fine, but I don't think it's a feasible approach in C/C++ without doing any strange hacks. The simplest way to overcome this problem is to store some kind of "context" which would allow the function to get back to work at the proper place. @tklengyel what do you think?

**Comment by tklengyel** [comment]
Keeping context should be doable, after all each plugin can keep track of the things they need internally. I'm trying to keep `libdrakvuf` as dumb as possible in that regard, so that it's only responsible for issuing the callbacks for all subscribers. Those subscribers can construct their own interpretation of the execution state.

**Comment by icedevml** [comment]
Return hooks are somehow bugged - this was probably fixed by #708

**Comment by icedevml** [comment]
As most of this stuff was already implemented/fixed => close the discussion
